src/ray/core_worker/reference_count.h

```diff
@@ -51,9 +79,41 @@ class ReferenceCounter {
   void LogDebugString() const LOCKS_EXCLUDED(mutex_);
 
  private:
+  /// Metadata for an ObjectID reference in the language frontend.
+  struct Reference {
+    /// Constructor for a reference whose origin is unknown.
+    Reference() : owned_by_us(false) {}
+    /// Constructor for a reference that we created.
+    Reference(const TaskID &owner_id, const rpc::Address &owner_address,
+              std::shared_ptr<std::vector<ObjectID>> deps)
+        : dependencies(std::move(deps)),
+          owned_by_us(true),
+          owner({owner_id, owner_address}) {}
+    /// Constructor for a reference that was given to us.
+    Reference(const TaskID &owner_id, const rpc::Address &owner_address)
```

Comment:
This is not quite a big deal, but it might be nice to have a `OwnedReference` and `BorrowedReference` , right now `dependencies` is used to differentiate the two cases, which doesn't seem very clear and might potentially lead to bugs.

Reply:
Actually the `owned_by_us` flag differentiates these cases.

Reply:
that's true, but the issue is that the flag is set internally in `Reference` class, and from the constructors it's not clear which one indicates `owned` and which one indicates `borrowed`. anyway it's not a big deal though so ok.